packages/components/src/navigator/navigator-provider/component.tsx

```diff
@@ -37,110 +37,86 @@ type RouterAction =
 	| { type: 'gotoparent'; options?: NavigateToParentOptions };
 
 type RouterState = {
+	initialPath: string;
 	screens: Screen[];
-	locationHistory: NavigatorLocation[];
+	currentLocation?: NavigatorLocation;
 	matchedPath: MatchedPath;
+	focusSelectors: Map< string, string >;
 };
 
-const MAX_HISTORY_LENGTH = 50;
-
 function addScreen( { screens }: RouterState, screen: Screen ) {
+	if ( screens.some( ( s ) => s.path === screen.path ) ) {
+		// eslint-disable-next-line no-console
+		console.warn(
+			`Navigator: a screen with path ${ screen.path } already exists.
+The screen with id ${ screen.id } will not be added.`
+		);
+		return screens;
+	}
 	return [ ...screens, screen ];
 }
 
 function removeScreen( { screens }: RouterState, screen: Screen ) {
 	return screens.filter( ( s ) => s.id !== screen.id );
 }
 
-function goBack( { locationHistory }: RouterState ) {
-	if ( locationHistory.length <= 1 ) {
-		return locationHistory;
-	}
-	return [
-		...locationHistory.slice( 0, -2 ),
-		{
-			...locationHistory[ locationHistory.length - 2 ],
-			isBack: true,
-			hasRestoredFocus: false,
-		},
-	];
-}
-
 function goTo(
 	state: RouterState,
 	path: string,
 	options: NavigateOptions = {}
 ) {
-	const { locationHistory } = state;
+	const { currentLocation, focusSelectors } = state;
+
 	const {
-		focusTargetSelector,
+		// Default assignments
 		isBack = false,
 		skipFocus = false,
-		replace = false,
+		// Extract to avoid forwarding
+		replace,
+		focusTargetSelector,
+		// Rest
 		...restOptions
 	} = options;
 
-	const isNavigatingToSamePath =
-		locationHistory.length > 0 &&
-		locationHistory[ locationHistory.length - 1 ].path === path;
-	if ( isNavigatingToSamePath ) {
-		return locationHistory;
+	if ( currentLocation?.path === path ) {
+		return currentLocation;
 	}
 
-	const isNavigatingToPreviousPath =
-		isBack &&
-		locationHistory.length > 1 &&
-		locationHistory[ locationHistory.length - 2 ].path === path;
+	// Set a focus selector that will be used when navigating
+	// back to the current location.
+	if ( focusTargetSelector && currentLocation?.path ) {
+		focusSelectors.set( currentLocation.path, focusTargetSelector );
+	}
 
-	if ( isNavigatingToPreviousPath ) {
-		return goBack( state );
+	// Get the focus selector for the new location.
+	let currentFocusSelector;
+	if ( isBack ) {
+		currentFocusSelector = focusSelectors.get( path );
+		// focusSelectors.delete( path );
```

Comment:
For some reason calling `delete` here would cause the location retrieved by the `NavigatorScreen` to have an undefined `focusTargetSelector`, as if `currentFocusSelector` was only holding a reference (and not the value) 

I was testing it and focus restoration in global styles didn't work until I commented out the line 🤷 

Would love it if someone could help clarify as this is unexpected to me

Reply:
🤦  I had somehow misset the lack of state update in the reducer, thank you for the help with the debugging.

Should be fixed in https://github.com/WordPress/gutenberg/pull/64675/commits/77fb59f4baa27133122a0320d4db07067cd3a5b8